### PR TITLE
Add dynamic config logic

### PR DESCRIPTION
**Pull request by roeiK-wix**
This is for loading a dynamic configuration if exists.
The dynamic config is a regular activity (probe) in the configuration section.
If there's a use-case for setting a configuration dynamically right before the experiment is starting.
It executes the probe,
and then the return value of this probe will be the config you wish to set.
The dictionary needs to have a key named `type` and as a value `probe`
alongside the rest of the probe props.
(No need for the `tolerance` key).

For example:
```json
    "some_dynamic_config": {
      "name": "some config probe",
      "type": "probe",
      "provider": {
        "type": "python",
        "module": "src.probes",
        "func": "config_probe",
        "arguments": {
            "arg1":"arg1"
        }
      }
    }
 ```
some_dynamic_config will be set with the return value of the function config_probe.

Side Note: the probe type can be the same as
a regular probe can be, python, os, etc...

The config argument is the configurations with all
the env vars configs that are already set.
(So basically we can use the configuration that
are injected from the environment in the config_probe arguments).

The secrets argument it's in case we need the secrets inside the config_probe.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,9 +1,22 @@
 # Changelog
 
 ## [Unreleased][]
+## [1.24.0][] - 2021-11-08
```
Comment (Lawouach) [inline CHANGELOG.md:4]: Do not set the version on PR :)

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,9 +1,22 @@
...
+
+
+[Unreleased]: https://github.com/chaostoolkit/chaostoolkit-lib/compare/1.24.0...HEAD
```
Comment (Lawouach) [inline CHANGELOG.md:7]: likewise

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,9 +1,22 @@
...
+
+### Added
+
+* `configuration.py` now have the ability to have load a dynamic configurations.
+  The dynamic config is a regular activity (probe) in the configuration section.
+  If there's a use-case for setting a configuration dynamically right before the experiment is starting.
+  It executes the probe,
+  and then the return value of this probe will be the config you wish to set.
```
Comment (Lawouach) [inline CHANGELOG.md:15]: This should be below the [Unreleased] on line 17

File: chaoslib/configuration.py [hunk 4]
```diff
@@ -71,7 +72,66 @@ def load_configuration(
...
+
+    Side Note: the probe type can be the same as
+    a regular probe can be, python, os, etc...
+
+    The config argument is the configurations with all
+    the env vars configs that are already set.
+    (So basically we can use the configuration that
+    are injected from the environment in the config_probe arguments).
+
+    The secrets argument it's in case we need the secrets inside the config_probe.
```
Comment (Lawouach) [inline chaoslib/configuration.py:124]: Could you rephrase this. I don't find it quite clear enough.

File: chaoslib/configuration.py [hunk 5]
```diff
@@ -71,7 +72,66 @@ def load_configuration(
...
+      "type": "probe",
+      "provider": {
+        "type": "python",
+        "module": "src.probes",
+        "func": "config_probe",
+        "arguments": {
+            "arg1":"arg1"
+        }
+      }
+    }
```
Comment (Lawouach) [inline chaoslib/configuration.py:111]: Set this between backticks please.

File: chaoslib/run.py [hunk 6]
```diff
@@ -278,6 +278,7 @@ def configure(
         self.secrets = load_secrets(
             experiment.get("secrets", {}), self.config, secret_vars
         )
+        self.config = load_dynamic_configuration(self.config, self.secrets)
```
Comment (Lawouach) [inline chaoslib/run.py:281]: Why is this a different call? Can't we call this from `load_configuration` itself?

**Review by Lawouach — CHANGES_REQUESTED** [review]
Starting the discussion. I would also appreciate some tests please.

**Comment by Lawouach** [comment]
Please rebase this branch from head of master and push again.

File: chaoslib/run.py (hunk 6 above)
Comment (roeiK-wix) [reply chaoslib/run.py:281]: The reason I created another function is because I had to pass to the function the secrets of the experiment, but the problem is that it's being created only after.
(Initially, I created this functionality inside the load_config function).

Comment (roeiK-wix) [reply chaoslib/run.py:281]: And another reason that it's after the regular load_config function is to get all the env vars after they are set in the load_config.

File: chaoslib/configuration.py [hunk 7]
```diff
@@ -3,6 +3,7 @@
 
 from logzero import logger
 
+from chaoslib.activity import run_activity
```
Comment (Lawouach) [inline chaoslib/configuration.py:6]: Would you mind trying to do that import from within the function and run `black` (our linter) again? The reason is that, with that import we bring a lot of other things inside a light module (configuration.py). Postponing the import may avoid that pity until needed.

File: chaoslib/configuration.py [hunk 8]
```diff
@@ -71,7 +72,68 @@ def load_configuration(
                         " that does not exist: {}".format(env_key)
                     )
                 conf[key] = extra_vars.get(key, env.get(env_key, env_default))
+            else:
+                conf[key] = extra_vars.get(key, value)
+
```
Comment (Lawouach) [inline chaoslib/configuration.py:77]: can you explain why this is added?

File: chaoslib/configuration.py [hunk 9]
```diff
@@ -71,7 +72,68 @@ def load_configuration(
...
         else:
             conf[key] = extra_vars.get(key, value)
 
     return conf
+
+
+def load_dynamic_configuration(
+    config: Dict[str, Any], secrets: Dict[str, Dict[str, str]]
```
Comment (Lawouach) [inline chaoslib/configuration.py:85]: we have types `Configuration` and `Secrets` for these

File: chaoslib/configuration.py [hunk 10]
```diff
@@ -71,7 +72,68 @@ def load_configuration(
...
+    after they are set in load_configuration.
+
+    The secrets argument contains all the secrets of the experiment.
+    """
+    conf = {}
+
+    logger.debug("Loading dynamic configuration...")
+    for (key, value) in config.items():
+        if isinstance(value, dict) and "type" in value:
+            if value["type"] == "probe":
```
Comment (Lawouach) [inline chaoslib/configuration.py:133]: These two lines could be squashed into one:

```
if isinstance(value, dict) and value.get("type") == "probe":
```

File: chaoslib/configuration.py [hunk 11]
```diff
@@ -71,7 +72,68 @@ def load_configuration(
...
+                value["provider"]["secrets"] = secrets
```
Comment (Lawouach) [inline chaoslib/configuration.py:134]: why?

**Review by Lawouach — CHANGES_REQUESTED** [review]
Sorry for the delay.

A few more comments.

File: chaoslib/configuration.py (hunk 8 above)
Comment (roeiK-wix) [reply chaoslib/configuration.py:88]: So if let says, I have two configurations, one with type env and the other with type probe (the dynamic config), the dynamic config will not appear and the `conf[key]` and so not appear in the self.config, and later on im passing the self.config to the load_dynamic_config.

**Review by Lawouach — APPROVED** [review]
LGTM
